--- a/src/gcc/ada/libgnat/s-os_lib.ads
+++ b/src/gcc/ada/libgnat/s-os_lib.ads
@@ -169,18 +169,11 @@ package System.OS_Lib is
    --  use OS_Time.
 
    subtype time_t is Long_Long_Integer;
-   --  C time_t can be either long or long long, so we choose the Ada
-   --  equivalent of the latter because eventually that will be the
-   --  type used out of necessity. This may affect some user code on 32-bit
-   --  targets that have not yet migrated to the Posix 2008 standard,
-   --  particularly pre version 5 32-bit Linux. Do not change this
-   --  declaration without coordinating it with conversions in Ada.Calendar.
+   --  Supported for backward compatibility, but no longer related
+   --  with the C time_t type.  For that, see System.C_Time.
 
    function To_C (Time : OS_Time) return time_t;
-   --  Convert OS_Time to C time_t type
-
    function To_Ada (Time : time_t) return OS_Time;
-   --  Convert C time_t type to OS_Time
 
    ----------------
    -- File Stuff --
